DM-53176: Add get_many_datasets()

dhirving opened the pull request:
Added a method `Butler.get_many_datasets()` to look up multiple `DatasetRef` instances from their UUIDs.  This will be used by the [DMTN-330](https://dmtn-330.lsst.io/) Prompt Publication Service when transferring groups of datasets between repositories.  

Also added support for passing dataset UUIDs as strings into the existing `get_dataset()` method, instead of requiring them to be manually converted to UUID objects.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Make get_dataset() accept hex strings

Often, users are calling get_dataset() because they have a dataset ID obtained from another source (copy/pasted from the CLI, etc.)  Usability is a lot better if they don't have to manually import uuid.UUID and construct an instance of it before calling get_dataset().

dhirving committed:
Fix type annotations for StaticDatasetTablesTuple

Fix type annotations for StaticDatasetTablesTuple so that typechecking and auto-complete work.  Previously this class was used to annotate two separate types, neither of which was fully specified.

dhirving committed:
De-optimize getDatasetRef's DatasetType lookup

Removed an optimization in getDatasetRef that allows it to avoid loading all dataset types from the database.  This optimization will be difficult to maintain in an upcoming commit to support fetching many datasets at the same time.

It would be unusual for an application to call get_dataset() and nothing else -- and even then, it only pays the cost of the fetch once, amid all the other Butler startup costs.  So it doesn't seem worth the cost of maintaining this extra cache update path.

dhirving committed:
Vectorize internals of getDataset

Rework the internals of getDataset to support looking up multiple refs at the
same time, in preparation for implementing get_many_datasets().

dhirving committed:
Remove unused DatasetTypeCache methods

dhirving committed:
Add Butler.get_many_datasets()

dhirving committed:
Remove loops over getDataset/get_dataset

Replace loops over getDataset with an equivalent call to get_many_datasets().

TallJimbo commented on python/lsst/daf/butler/registry/datasets/byDimensions/_manager.py:
```diff
@@ -422,50 +423,63 @@ def resolve_wildcard(
+            id_col,
+            self._static.dataset.columns["dataset_type_id"],
+            run_key_column,
+        ).where(id_col.in_(ids))
```
If you add `.subquery()` to the end of this, you'll get an equivalent `FromClause` to pass to `lookup_name_sql`, and you can drop the `Joinable` type alias and all of the annotation changes that use it (which otherwise would need some docstring updates to go along with them).

TallJimbo replied:
Did you miss this comment?

dhirving replied:
I adjusted the docstrings.  I don't think it's worth subjecting ourselves to the vagaries of Postgres' subquery optimization logic to avoid having a type definition.

TallJimbo replied:
I didn't think there would be any difference in the SQL that would be generated; I was thinking it was a question of whether you add the `subquery()` or SQLAlchemy does it internally.  But maybe I'm wrong about that.

dhirving replied:
According to the `SelectBase.subquery()` docs it renders as a parenthesized subquery:

```
        The subquery form by itself renders the same way, however when
        embedded into the FROM clause of another SELECT statement, it becomes
        a named sub-element::

            subq = stmt.subquery()
            new_stmt = select(subq)

        The above renders as:

        .. sourcecode:: sql

            SELECT anon_1.id, anon_1.name
            FROM (SELECT table.id, table.name FROM table) AS anon_1
```

Whether that's true in all cases I'm not sure, but during development I did confirm that without the `subquery()` it generates a totally normal `SELECT ... FROM ... INNER JOIN ...` query.

TallJimbo commented on python/lsst/daf/butler/registry/datasets/byDimensions/_manager.py:
```diff
@@ -422,50 +423,63 @@ def resolve_wildcard(
+                # In theory the data ID for a given dataset should be the
+                # same in both, but nothing actually guarantees this.
+                # So skip any tagged collections, using the run collection
+                # as the definitive definition.
```
Seems like it might be better to join to the collections table and require the collection type there to be RUN, especially if someday we have a lot of TAGGED collections (as I think we're generally planning to use them more in the future than we are now).

dhirving committed:
Filter tag collections in DB instead of application

In response to Jim's review comment, join the collections table to filter out the excess rows from tagged collections instead fetching all the rows and filtering in the application.  This will prevent us from loading an excess number of rows for datasets that are tagged in many collections, and should have minimal performance impact because we have to join to get the collection name in either case.
